### ISSUE TITLE
inconsistent go vendoring

### ISSUE THREAD

**Issue by johanneskastl** [ChaosInTheCRD/kube-lock#5]
Could you please check the vendoring information in the go.mod file (by running `go mod tidy` if I remember correctly)?

I get the following error:

```
[   47s] + go build -mod=vendor -buildmode=pie -o bin/kube-lock .                                                                                                                            
[   47s] go: inconsistent vendoring in /home/abuild/rpmbuild/BUILD/kube-lock-0.0.4:                                                                                                          
[   47s]        github.com/gertd/go-pluralize@v0.2.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/manifoldco/promptui@v0.9.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/sirupsen/logrus@v1.9.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/spf13/cobra@v1.6.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/spf13/viper@v1.15.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        gopkg.in/yaml.v3@v3.0.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        k8s.io/client-go@v0.26.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt                              
[   47s]        github.com/chzyer/readline@v1.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt              
[   47s]        github.com/davecgh/go-spew@v1.1.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt                  
[   47s]        github.com/emicklei/go-restful/v3@v3.10.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt  
[   47s]        github.com/fsnotify/fsnotify@v1.6.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/go-logr/logr@v1.2.3: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/go-openapi/jsonpointer@v0.19.6: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/go-openapi/jsonreference@v0.20.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/go-openapi/swag@v0.22.3: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/gogo/protobuf@v1.3.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/golang/protobuf@v1.5.3: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/google/btree@v1.1.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/google/gnostic@v0.6.9: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/google/go-cmp@v0.5.9: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/google/gofuzz@v1.2.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/google/uuid@v1.3.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/gregjones/httpcache@v0.0.0-20190611155906-901d90724c79: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/hashicorp/hcl@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/imdario/mergo@v0.3.14: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/inconshreveable/mousetrap@v1.1.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/josharian/intern@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/json-iterator/go@v1.1.12: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/magiconair/properties@v1.8.7: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/mailru/easyjson@v0.7.7: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/mitchellh/mapstructure@v1.5.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/modern-go/concurrent@v0.0.0-20180306012644-bacd9c7ef1dd: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/modern-go/reflect2@v1.0.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/munnerz/goautoneg@v0.0.0-20191010083416-a7dc8b61c822: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/pelletier/go-toml/v2@v2.0.7: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/peterbourgon/diskv@v2.0.1+incompatible: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/spf13/afero@v1.9.5: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/spf13/cast@v1.5.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/spf13/jwalterweatherman@v1.1.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/spf13/pflag@v1.0.5: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/stretchr/testify@v1.8.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        github.com/subosito/gotenv@v1.4.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        golang.org/x/net@v0.8.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        golang.org/x/oauth2@v0.6.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        golang.org/x/sys@v0.6.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        golang.org/x/term@v0.6.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        golang.org/x/text@v0.8.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        golang.org/x/time@v0.3.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        google.golang.org/appengine@v1.6.7: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        google.golang.org/protobuf@v1.30.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        gopkg.in/inf.v0@v0.9.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        gopkg.in/ini.v1@v1.67.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        gopkg.in/yaml.v2@v2.4.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        k8s.io/api@v0.26.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        k8s.io/apimachinery@v0.26.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        k8s.io/klog/v2@v2.90.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        k8s.io/kube-openapi@v0.0.0-20230308215209-15aac26d736a: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        k8s.io/utils@v0.0.0-20230313181309-38a27ef9d749: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        sigs.k8s.io/json@v0.0.0-20221116044647-bc3834ca7abd: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        sigs.k8s.io/structured-merge-diff/v4@v4.2.3: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
[   47s]        sigs.k8s.io/yaml@v1.3.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
```

**Comment by ChaosInTheCRD** [comment]
Hi again 👋

It looks like what you are looking for can be achieved with `go vendor`. However, may I ask why you need this? I am afraid that if I vendor all dependencies into the repo then I will have to keep them up to date every time I make a change (executing `go vendor`) and I am not sure of any other drawbacks to doing this. Could you not build it by removing the `-mod=vendor` flag?

**Comment by johanneskastl** [comment]
No need to keep anything in the repository. But the go.mod file should be consistent.

In my case I am preparing a package for openSUSE, where builds are only done offline. So I need to vendor everything up front. Which is what is failing due to inconsistencies in the go.mod or go.sum files.

I am not a go expert, but this far I was told that running `go mod tidy` seems to be necessary for cleaning up the files.

**Comment by johanneskastl** [comment]
https://go.dev/doc/modules/managing-dependencies

> Synchronizing your code’s dependencies
> 
> You can ensure that you’re managing dependencies for all of your code’s imported packages while also removing dependencies for packages you’re no longer importing.
> 
> This can be useful when you’ve been making changes to your code and dependencies, possibly creating a collection of managed dependencies and downloaded modules that no longer match the collection specifically required by the packages imported in your code.
> 
> To keep your managed dependency set tidy, use the go mod tidy command. Using the set of packages imported in your code, this command edits your go.mod file to add modules that are necessary but missing. It also removes unused modules that don’t provide any relevant packages.
> 
> The command has no arguments except for one flag, -v, that prints information about removed modules.

**Comment by ChaosInTheCRD** [comment]
Sorry for the slow response!

I did some exploration and asked around about this a while back. The command that you referenced above is using `-mod=vendor` to perform the build. This is what is leading to the errors below as `go vendor` hasn't been called to create a `vendor/modules.txt`.

I am pretty sure that you can perform this before the build for the OpenSUSE package, but I don't think it makes sense to start vendoring all dependencies in the project upstream.

I am going to close this issue as I think I have adequately responded 😄. Hopefully my answer makes sense.

**Comment by johanneskastl** [comment]
Have you tried what happens when you run `go mod tidy`? In my understanding this does not keep anything in the repository, but it make the file consistent.

> I am pretty sure that you can perform this before the build for the OpenSUSE package

This is what I am trying, but this does not work due to the inconsistent vendoring.

**Comment by ChaosInTheCRD** [comment]
Hey 👋 

```
kube-lock on main via 🐹 v1.21.4 
➜ go mod tidy        

kube-lock on main via 🐹 v1.21.4 
➜ git --no-pager diff
```

The above gleans no output as the `go.mod` and `go.sum` which is responsible for pinning the dependencies on the project is up to date.

In the log output you referenced above you are executing:
```
go build -mod=vendor -buildmode=pie -o bin/kube-lock .   
```

This gives the following output:
```
kube-lock on  main via 🐹 v1.21.4 using ☁️  default/mustang-295918 
➜ go build -mod=vendor -buildmode=pie -o bin/kube-lock .   
go: inconsistent vendoring in /Users/chaosinthecrd/Git/kube-lock:
	github.com/manifoldco/promptui@v0.9.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/sirupsen/logrus@v1.9.3: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/spf13/cobra@v1.8.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/spf13/viper@v1.18.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	gopkg.in/yaml.v3@v3.0.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	k8s.io/client-go@v0.29.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
...
	To ignore the vendor directory, use -mod=readonly or -mod=mod.
	To sync the vendor directory, run:
		go mod vendor
```

Which is inline with what you experienced. Notice at the bottom it's giving options as to what you can do.
- ignore the vendor directory (by not using `-mod=vendor`), and instead use `-mod=readonly` or `-mod=mod`.
- sync the vendor directory by running `go mod vendor`.

In your case, you should be fine to just execute `go mod vendor` in your pipeline if you need a vendor directory:
```
kube-lock on main via 🐹 v1.21.4
➜ go mod vendor                                             

kube-lock on main [?] via 🐹 v1.21.4
➜ git status         
On branch main
Your branch is up to date with 'origin/main'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	vendor/
```

Notice that there is now a `vendor/` directory in the repository.

Now, if I go ahead and build with the parameters you specified:
```
kube-lock on main [?] via 🐹 v1.21.4
➜ go build -mod=vendor -buildmode=pie -o bin/kube-lock .  

kube-lock on main [?] via 🐹 v1.21.4
➜ tree bin 
bin
└── kube-lock

1 directory, 1 file
```

I now have a build binary.

TLDR:
- Either build the tool without vendoring enabled:
```
kube-lock on main [?] via 🐹 v1.21.4 
➜ go build -mod=mod -buildmode=pie -o bin/kube-lock . 
```
- Or vendor the dependencies with `go mod vendor` and find a way of doing that so your pipeline can utilise it when calling `go build -mod=vendor`.

I don't wish to vendor the dependencies in the repository as it's an extra thing to manage, and the `go.mod` is sufficient. If this is really necessary and there is a high level of demand for it though, I would consider changing my stance.

Hope this helps.

Update:

Apologies, I recognise that this isn't totally helpful when answering your question:

> In my case I am preparing a package for openSUSE, where builds are only done offline. So I need to vendor everything up front. Which is what is failing due to inconsistencies in the go.mod or go.sum files.

One option that we could explore is https://github.com/marketplace/actions/rpm-build ([OpenSUSE supports RPM](https://documentation.suse.com/sbp/systems-management/pdf/SBP-RPM-Packaging_en.pdf), would this fit the bill for what you need?

**Comment by johanneskastl** [comment]
Ha! In your face! Sorry, talking to myself.

I finally had time to dig deeper and it turns out this was all due to a stupid mistake. The vendor dependencies are fine, they were vendored just fine but not placed at the proper place during the build. Which meant they were not found...

Sorry for the noise, and thanks for your help and understanding.

Kind Regards
Johannes

P.S. In other words, there are now working package on their way to openSUSE Tumbleweed...
